src/workerd/api/streams/internal.c++

```diff
@@ -90,22 +90,24 @@ private:
     size_t amountToRead = kj::min(MAX_BUFFER_CHUNK,
         input.tryGetLength(StreamEncoding::IDENTITY).orDefault(DEFAULT_BUFFER_CHUNK));
 
-    for (;;) {
-      // TODO(perf): We can likely further optimize this loop by checking to see
-      // how much of the buffer was filled and using the remaining buffer space if
-      // it is not completely filled by the previous iteration. Doing so makes this
-      // loop a bit more complicated tho, so for now let's keep things simple.
-      auto bytes = kj::heapArray<T>(amountToRead);
-      size_t amount = co_await input.tryRead(bytes.begin(), 1, bytes.size());
-
-      if (amount == 0) {
-        break;
-      }
+    if (amountToRead > 0) {
```

Comment:
This fixes the correctness bug.

I think there's also some performance bugs, which could be fixed in a later PR:
* It looks like when we chose `amountToRead` to be exactly the stream length above, then we'll always additionally allocate a second buffer of this size, only to discover that there are no bytes available to read into it.
* If the `tryRead()` below only fills part of the buffer, we nevertheless place it into the `parts` list and allocate a new buffer for the next read. This could be extremely memory-inefficient for many short reads. We should really try to fill one buffer entirely before allocating the next one.

Reply:
I see the first issue is newly-introduced in #1137. The second issue is technically not new, but could become worse if we're allocating larger buffers now due to #1137.

Maybe we should revert #1137 until it can be more thoroughly tested and reviewed?

Reply:
Notice that there's already a TODO comment indicating this :-) 

```
      // TODO(perf): We can likely further optimize this loop by checking to see
      // how much of the buffer was filled and using the remaining buffer space if
      // it is not completely filled by the previous iteration. Doing so makes this
      // loop a bit more complicated tho, so for now let's keep things simple.
```

I already have the more optimized version of this ready locally but since we're holding off landing non-critical fixes in main I haven't opened the PR yet.

Reply:
If you'd feel more comfortable reverting 1137 and just landing the more optimized version later, then no objections.

Reply:
That sounds best to me but it's up to you and @Warfields.

Reply:
I'd rather not have more pile onto the backlog of PRs.